### PR TITLE
HPA: Don't mutate the shared informer cache

### LINKED ISSUE #41768: flake: podautoscaler unit test failing
Just a few:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41149/pull-kubernetes-unit/18979/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41077/pull-kubernetes-unit/18972/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41714/pull-kubernetes-unit/18970/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/34259/pull-kubernetes-unit/18968/

Many more:
https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-unit

Couldn't recreate locally.

### PR DESCRIPTION
**Pull request by DirectXMan12**
Conversions can mutate the underlying object (and ours were).
Make a deepcopy before our first conversion at the very start
of the reconciler method in order to avoid mutating the shared
informer cache during conversion.

Fixes #41768 

```release-note
NONE
```

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @ncdc @smarterclayton

File: pkg/controller/podautoscaler/horizontal.go [hunk 1]
```diff
@@ -58,8 +58,8 @@ func calculateScaleUpLimit(currentReplicas int32) int32 {
 
 // ConvertToVersionVia is like api.Scheme.ConvertToVersion, but it does so via an internal version first.
```
Comment (ncdc) [inline pkg/controller/podautoscaler/horizontal.go]: UnsafeConvertToVersionVia

File: pkg/controller/podautoscaler/horizontal.go [hunk 2]
```diff
@@ -58,8 +58,8 @@ func calculateScaleUpLimit(currentReplicas int32) int32 {
...
 // We use it since working with v2alpha1 is convinient here, but we want to use the v1 client (and
```
Comment (ncdc) [inline pkg/controller/podautoscaler/horizontal.go]: convenient

File: pkg/controller/podautoscaler/horizontal.go [hunk 3]
```diff
@@ -275,8 +275,16 @@ func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.Hori
...
 }
 
-func (a *HorizontalController) reconcileAutoscaler(hpav1 *autoscalingv1.HorizontalPodAutoscaler) error {
-	// first, convert to autoscaling/v2, which makes our lives easier when calculating metrics
+func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.HorizontalPodAutoscaler) error {
+	// make a copy so that we never mutate the shared informer cache (conversion can mutate the object)
+	hpav1Raw, err := api.Scheme.DeepCopy(hpav1Shared)
+	if err != nil {
+		a.eventRecorder.Event(hpav1Shared, v1.EventTypeWarning, "FailedConvertHPA", err.Error())
+		return fmt.Errorf("failed to deecopy the HPA: %v", err)
```
Comment (ncdc) [inline pkg/controller/podautoscaler/horizontal.go]: typo: deep copy

**Comment by ncdc** [comment]
cc @wojtek-t

**Comment by deads2k** [comment]
@k8s-bot kubemark e2e test this

File: pkg/controller/podautoscaler/horizontal.go [hunk 4]
```diff
@@ -56,10 +56,10 @@ func calculateScaleUpLimit(currentReplicas int32) int32 {
...
 
-// ConvertToVersionVia is like api.Scheme.ConvertToVersion, but it does so via an internal version first.
-// We use it since working with v2alpha1 is convinient here, but we want to use the v1 client (and
-// can't just use the internal version).  Note that it does *not* guarantee a copy is made -- this should
-// be done separately if we need to mutate the object.
+// UnsafeConvertToVersionVia is like api.Scheme.UnsafeConvertToVersion, but it does so via an internal version first.
+// We use it since working with v2alpha1 is convenient here, but we want to use the v1 client (and
+// can't just use the internal version).  Note that conversion mutates the object, so you need to deepcopy
+// *before* you call this if the input object came out of a shared cache.
 func UnsafeConvertToVersionVia(obj runtime.Object, externalVersion schema.GroupVersion) (runtime.Object, error) {
```
Comment (deads2k) [inline pkg/controller/podautoscaler/horizontal.go]: isn't there a spot in the tests that calls this?  I don't remember seeing it deep copy before calling this.  Why doesn't it need to?

**Comment by deads2k** [comment]
question, lgtm otherwise.

**Comment by ncdc** [comment]
@k8s-bot gce etcd3 e2e test this

File: pkg/controller/podautoscaler/horizontal.go (hunk 4 above)
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/horizontal.go]: because it's an object we created 10 lines up that doesn't escape the scope

**Comment by ncdc** [comment]
Green. Ready to tag @deads2k ?

**Comment by ncdc** [comment]
@k8s-oncall we probably need to raise the priority on this. Because of the cache mutation detector that's now enabled on some PR test jobs, the kube-controller-manager is going to panic and terminate as soon as the HPA controller starts to process an HPA.

cc @smarterclayton

**Comment by wojtek-t** [comment]
@k8s-bot etcd3 test this, issue: #IGNORE

**Comment by wojtek-t** [comment]
Those flakes are so painful that I'm merging this PR manually (the tests are now green).

**Comment by ncdc** [comment]
Thanks!

**Pull request merged**